File: android/app/src/main/kotlin/com/aw/huda/widget/prayer/PrayerWidgetCalculator.kt
package com.aw.huda.widget.prayer

import io.github.abdulwaheds.prayertimeplus.AutoMethod
import io.github.abdulwaheds.prayertimeplus.CalculationMethod
import io.github.abdulwaheds.prayertimeplus.Coordinates
import io.github.abdulwaheds.prayertimeplus.DateComponents
import io.github.abdulwaheds.prayertimeplus.HighLatitudeRule
import io.github.abdulwaheds.prayertimeplus.Madhab
import io.github.abdulwaheds.prayertimeplus.PrayerTimes
import io.github.abdulwaheds.prayertimeplus.SunnahTimes
import java.time.OffsetDateTime
import java.time.LocalDate
import java.time.LocalTime
import java.time.chrono.HijrahDate
import java.time.temporal.ChronoField
import java.util.Calendar
import java.util.Date
import kotlin.math.ceil

internal object PrayerWidgetCalculator {
    private const val DEFAULT_FUTURE_DAYS = 8
    private const val MAX_OFFSET_MINUTES = 7 * 24 * 60

    data class DayTimes(
        val civilDate: DateComponents,
        val fajr: Date?,
        val sunrise: Date?,
        val dhuhr: Date?,
        val asr: Date?,
        val maghrib: Date?,
        val isha: Date?,
    ) {
        val ordered: List<Pair<PrayerKind, Date>>
            get() = listOfNotNull(
                fajr?.let { PrayerKind.FAJR to it },
                dhuhr?.let { PrayerKind.DHUHR to it },
                asr?.let { PrayerKind.ASR to it },
                maghrib?.let { PrayerKind.MAGHRIB to it },
                isha?.let { PrayerKind.ISHA to it },
            )

        fun timeOf(kind: PrayerKind): Date? = when (kind) {
            PrayerKind.FAJR -> fajr
            PrayerKind.SUNRISE -> sunrise
            PrayerKind.DHUHR -> dhuhr
            PrayerKind.ASR -> asr
            PrayerKind.MAGHRIB -> maghrib
            PrayerKind.ISHA -> isha
        }
    }

    data class PrayerEvent(
        val kind: PrayerKind,
        val time: Date,
        val day: DayTimes,
    )

    fun computeDay(snapshot: PrayerWidgetSnapshot, day: Calendar): DayTimes? {
        val raw = rawTimes(snapshot, day) ?: return null
        val date = DateComponents(
            day.get(Calendar.YEAR),
            day.get(Calendar.MONTH) + 1,
            day.get(Calendar.DAY_OF_MONTH),
        )
        return rawToDay(raw, snapshot, date)
    }

    private fun rawTimes(snapshot: PrayerWidgetSnapshot, day: Calendar): PrayerTimes? {
        val lat = snapshot.latitude ?: return null
        val lon = snapshot.longitude ?: return null
        val countryCode = snapshot.countryCode?.trim().orEmpty()
        val method = methodFrom(snapshot.calculationMethod, countryCode)
        val isRamadan = isRamadan(snapshot, day)
        var params = method.parameters().copy(
            madhab = madhabFrom(snapshot.madhab),
            highLatitudeRule = highLatitudeRuleFrom(snapshot.highLatitudeRule),
            isRamadan = isRamadan,
        )
        if (method == CalculationMethod.UMM_AL_QURA &&
            isRamadan && countryCode.uppercase() != "SA"
        ) {
            params = params.copy(ishaValue = 120.0)
        }
        if (method == CalculationMethod.OTHER) {
            params = params.copy(
                fajrAngle = snapshot.customFajrAngle,
                maghribIsInterval = false,
                maghribValue = snapshot.customMaghribAngle,
                ishaIsInterval = false,
                ishaValue = snapshot.customIshaAngle,
            )
        }

        val date = DateComponents(
            day.get(Calendar.YEAR),
            day.get(Calendar.MONTH) + 1,
            day.get(Calendar.DAY_OF_MONTH),
        )
        val utcOffset = LocalDate.of(date.year, date.month, date.day)
            .atTime(LocalTime.NOON)
            .atZone(snapshot.displayTimeZone.toZoneId())
            .offset
        return PrayerTimes(
            Coordinates(lat, lon),
            date,
            params,
            utcOffset,
            countryCode = countryCode,
        )
    }

    private fun rawToDay(
        raw: PrayerTimes,
        snapshot: PrayerWidgetSnapshot,
        civilDate: DateComponents,
    ): DayTimes? {
        val day = DayTimes(
            civilDate = civilDate,
            fajr = raw.fajr?.toDate()?.applyOffset(offset(snapshot, "fajr")),
            sunrise = raw.sunrise?.toDate()?.applyOffset(offset(snapshot, "sunrise")),
            dhuhr = raw.dhuhr?.toDate()?.applyOffset(offset(snapshot, "dhuhr")),
            asr = raw.asr?.toDate()?.applyOffset(offset(snapshot, "asr")),
            maghrib = raw.maghrib?.toDate()?.applyOffset(offset(snapshot, "maghrib")),
            isha = raw.isha?.toDate()?.applyOffset(offset(snapshot, "isha")),
        )
        return day.takeIf { displayList(it).isNotEmpty() }
    }

    fun computeSunnah(snapshot: PrayerWidgetSnapshot, day: Calendar): SunnahTimesResult? {
        val raw = rawTimes(snapshot, day) ?: return null
        return try {
            val sunnah = SunnahTimes(raw)
            val middle = sunnah.middleOfTheNight?.toDate() ?: return null
            val lastThird = sunnah.lastThirdOfTheNight?.toDate() ?: return null
            SunnahTimesResult(middleOfNight = middle, lastThirdOfNight = lastThird)
        } catch (_: Exception) {
            null
        }
    }

    fun displayList(day: DayTimes): List<Pair<PrayerKind, Date>> = listOfNotNull(
        day.fajr?.let { PrayerKind.FAJR to it },
        day.sunrise?.let { PrayerKind.SUNRISE to it },
        day.dhuhr?.let { PrayerKind.DHUHR to it },
        day.asr?.let { PrayerKind.ASR to it },
        day.maghrib?.let { PrayerKind.MAGHRIB to it },
        day.isha?.let { PrayerKind.ISHA to it },
    )

    fun eventTimeline(
        snapshot: PrayerWidgetSnapshot,
        around: Date,
        futureDays: Int = DEFAULT_FUTURE_DAYS,
    ): List<PrayerEvent> {
        if (!snapshot.hasCoordinates) return emptyList()
        val offsetDays = ceil(
            snapshot.offsets.keys.maxOfOrNull { kotlin.math.abs(offset(snapshot, it)) }
                .orZero() / (24.0 * 60.0),
        ).toInt()
        val anchor = calendarFor(snapshot, around)
        val events = ArrayList<PrayerEvent>()
        for (dayOffset in -(offsetDays + 2)..(futureDays + offsetDays)) {
            val dayCalendar = (anchor.clone() as Calendar).apply {
                add(Calendar.DATE, dayOffset)
            }
            val day = computeDay(snapshot, dayCalendar) ?: continue
            for ((kind, time) in day.ordered) {
                events += PrayerEvent(kind, time, day)
            }
        }
        return events.sortedWith(compareBy<PrayerEvent> { it.time.time }.thenBy { it.kind.ordinal })
    }

    fun nextAfter(snapshot: PrayerWidgetSnapshot, now: Date): NextPrayer? =
        eventTimeline(snapshot, now).firstOrNull { it.time.after(now) }
            ?.let { NextPrayer(it.kind, it.time, it.day) }

    fun previousBefore(snapshot: PrayerWidgetSnapshot, now: Date): PreviousPrayer? =
        eventTimeline(snapshot, now).lastOrNull { !it.time.after(now) }
            ?.let { PreviousPrayer(it.kind, it.time) }

    @Suppress("UNUSED_PARAMETER")
    private fun isRamadan(snapshot: PrayerWidgetSnapshot, day: Calendar): Boolean = try {
        val civilDate = LocalDate.of(
            day.get(Calendar.YEAR),
            day.get(Calendar.MONTH) + 1,
            day.get(Calendar.DAY_OF_MONTH),
        )
        HijrahDate.from(civilDate).get(ChronoField.MONTH_OF_YEAR) == 9
    } catch (_: Exception) {
        false
    }

    private fun OffsetDateTime.toDate(): Date = Date.from(toInstant())
    private fun madhabFrom(token: String?): Madhab =
        if (token == "hanafi") Madhab.HANAFI else Madhab.SHAFI

    private fun highLatitudeRuleFrom(token: String?): HighLatitudeRule = when (token) {
        "automatic" -> HighLatitudeRule.AUTOMATIC
        "middleOfTheNight" -> HighLatitudeRule.MIDDLE_OF_THE_NIGHT
        "seventhOfTheNight" -> HighLatitudeRule.SEVENTH_OF_THE_NIGHT
        "twilightAngle" -> HighLatitudeRule.TWILIGHT_ANGLE
        "none" -> HighLatitudeRule.NONE
        else -> HighLatitudeRule.AUTOMATIC
    }

    internal fun methodFrom(token: String?, countryCode: String): CalculationMethod {
        if (token.isNullOrBlank() || token == "auto") {
            return if (countryCode.isBlank()) CalculationMethod.UMM_AL_QURA
            else AutoMethod.forCountry(countryCode)
        }
        val methodKey = when (token) {
            "muslimWorldLeague" -> "mwl"
            "egyptian" -> "egypt"
            "ummAlQura" -> "makkah"
            "northAmerica" -> "isna"
            "southKorea" -> "southkorea"
            "other" -> "custom"
            else -> token
        }
        return CalculationMethod.fromKey(methodKey) ?: CalculationMethod.UMM_AL_QURA
    }

    private fun offset(snapshot: PrayerWidgetSnapshot, key: String): Int =
        (snapshot.offsets[key] ?: 0).coerceIn(-MAX_OFFSET_MINUTES, MAX_OFFSET_MINUTES)

    private fun Date.applyOffset(minutes: Int): Date =
        if (minutes == 0) this else Date(time + minutes * 60L * 1_000L)

    private fun calendarFor(snapshot: PrayerWidgetSnapshot, now: Date): Calendar =
        Calendar.getInstance(snapshot.displayTimeZone).apply { time = now }

    private fun Int?.orZero(): Int = this ?: 0
}

internal enum class PrayerKind { FAJR, SUNRISE, DHUHR, ASR, MAGHRIB, ISHA }

internal data class NextPrayer(
    val kind: PrayerKind,
    val time: Date,
    val day: PrayerWidgetCalculator.DayTimes,
)

internal data class PreviousPrayer(val kind: PrayerKind, val time: Date)
internal data class SunnahTimesResult(val middleOfNight: Date, val lastThirdOfNight: Date)

internal fun progressStart(previous: PreviousPrayer?, target: Date): Date {
    if (previous != null && previous.time.before(target)) return previous.time
    return Date(target.time - 6L * 60L * 60L * 1_000L)
}

File: ios/HudaWidget/PrayerWidgetCalculator.swift
import Foundation

struct PrayerWidgetDayTimes {
    let base: PrayerTimes
    let offsets: [Prayer: Int]
    let timeZone: TimeZone

    func time(for prayer: Prayer) -> Date? {
        guard let date = base.time(for: prayer) else { return nil }
        var fixedCalendar = Calendar(identifier: .gregorian)
        fixedCalendar.timeZone = TimeZone(
            secondsFromGMT: Int(base.utcOffset)
        ) ?? timeZone
        let components = fixedCalendar.dateComponents(
            [.year, .month, .day, .hour, .minute, .second],
            from: date
        )
        var locationCalendar = Calendar(identifier: .gregorian)
        locationCalendar.timeZone = timeZone
        guard let locationDate = locationCalendar.date(from: components) else {
            return nil
        }
        return PrayerWidgetCalculator.applyingManualOffset(
            to: locationDate,
            minutes: offsets[prayer] ?? 0
        )
    }

    var fajr: Date? {
        time(for: .fajr)
    }

    var maghrib: Date? {
        time(for: .maghrib)
    }
}

enum PrayerWidgetCalculator {
    static let dailyPrayers: [Prayer] = [.fajr, .dhuhr, .asr, .maghrib, .isha]

    static let displayPrayers: [Prayer] = [
        .fajr, .sunrise, .dhuhr, .asr, .maghrib, .isha,
    ]

    static func applyingManualOffset(to date: Date, minutes: Int) -> Date {
        date.addingTimeInterval(TimeInterval(minutes * 60))
    }

    static func computeTimes(
        coordinates: Coordinates,
        date: Date,
        settings: PrayerWidgetSettings
    ) -> PrayerWidgetDayTimes? {
        let countryCode = settings.countryCode.trimmingCharacters(in: .whitespaces)
        let calculationMethod = method(
            from: settings.calculationMethod,
            countryCode: countryCode
        )
        var params = calculationMethod.parameters
        params.madhab = madhab(from: settings.madhab)
        params.highLatitudeRule = highLatitudeRule(from: settings.highLatitudeRule)
        var islamicCalendar = Calendar(identifier: .islamicUmmAlQura)
        islamicCalendar.timeZone = settings.displayTimeZone
        let isRamadan = islamicCalendar.component(.month, from: date) == 9
        params.isRamadan = isRamadan
        if calculationMethod == .ummAlQura,
           isRamadan,
           countryCode.uppercased() != "SA"
        {
            params.ishaValue = 120
        }
        if calculationMethod == .other {
            params.fajrAngle = settings.customFajrAngle
            params.maghribIsInterval = false
            params.maghribValue = settings.customMaghribAngle
            params.ishaIsInterval = false
            params.ishaValue = settings.customIshaAngle
        }

        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = settings.displayTimeZone
        let components = calendar.dateComponents([.year, .month, .day], from: date)
        var noonComponents = components
        noonComponents.hour = 12
        let offsetReference = calendar.date(from: noonComponents) ?? date
        let base = PrayerTimes(
            coordinates: coordinates,
            date: components,
            calculationParameters: params,
            utcOffset: TimeInterval(
                settings.displayTimeZone.secondsFromGMT(for: offsetReference)
            ),
            countryCode: countryCode
        )
        return PrayerWidgetDayTimes(
            base: base,
            offsets: settings.offsets,
            timeZone: settings.displayTimeZone
        )
    }

    static func dailyMap(from times: PrayerWidgetDayTimes) -> [(Prayer, Date)] {
        dailyPrayers.compactMap { prayer in
            times.time(for: prayer).map { (prayer, $0) }
        }
    }

    static func displayMap(from times: PrayerWidgetDayTimes) -> [(Prayer, Date)] {
        displayPrayers.compactMap { prayer in
            times.time(for: prayer).map { (prayer, $0) }
        }
    }

    static func nextTransition(
        coordinates: Coordinates,
        settings: PrayerWidgetSettings,
        from now: Date
    ) -> (prayer: Prayer, date: Date)? {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = settings.displayTimeZone
        for dayOffset in 0 ..< 5 {
            guard
                let target = calendar.date(byAdding: .day, value: dayOffset, to: now),
                let times = computeTimes(coordinates: coordinates, date: target, settings: settings)
            else { continue }
            for (prayer, date) in dailyMap(from: times) where date > now {
                return (prayer, date)
            }
        }
        return nil
    }

    static func transitions(
        coordinates: Coordinates,
        settings: PrayerWidgetSettings,
        startingAt start: Date,
        dayCount: Int
    ) -> [(prayer: Prayer, date: Date)] {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = settings.displayTimeZone
        var result: [(Prayer, Date)] = []
        for dayOffset in 0 ..< dayCount {
            guard
                let target = calendar.date(byAdding: .day, value: dayOffset, to: start),
                let times = computeTimes(coordinates: coordinates, date: target, settings: settings)
            else { continue }
            for entry in dailyMap(from: times) {
                result.append(entry)
            }
        }
        return result.sorted { $0.1 < $1.1 }
    }

    static func sunnahTimes(
        coordinates: Coordinates,
        settings: PrayerWidgetSettings,
        date: Date
    ) -> (middle: Date?, lastThird: Date?) {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = settings.displayTimeZone
        guard
            let today = computeTimes(
                coordinates: coordinates,
                date: date,
                settings: settings
            ),
            let tomorrowDate = calendar.date(byAdding: .day, value: 1, to: date),
            let tomorrow = computeTimes(
                coordinates: coordinates,
                date: tomorrowDate,
                settings: settings
            ),
            let maghrib = today.maghrib,
            let nextFajr = tomorrow.fajr,
            nextFajr > maghrib
        else { return (nil, nil) }
        let night = nextFajr.timeIntervalSince(maghrib)
        func rounded(_ value: Date) -> Date {
            Date(timeIntervalSince1970: (value.timeIntervalSince1970 / 60).rounded() * 60)
        }
        return (
            rounded(maghrib.addingTimeInterval(night / 2)),
            rounded(maghrib.addingTimeInterval(night * 2 / 3))
        )
    }

    // MARK: Settings mapping

    static func method(from token: String, countryCode: String) -> CalculationMethod {
        if token.isEmpty || token == "auto" {
            return countryCode.isEmpty ? .ummAlQura : AutoMethod.forCountry(countryCode)
        }

        let methodKey: String
        switch token {
        case "muslimWorldLeague": methodKey = "mwl"
        case "egyptian": methodKey = "egypt"
        case "ummAlQura": methodKey = "makkah"
        case "northAmerica": methodKey = "isna"
        case "southKorea": methodKey = "southkorea"
        case "other": return .other
        default: methodKey = token
        }
        return CalculationMethod.from(key: methodKey) ?? .ummAlQura
    }

    private static func madhab(from token: String) -> Madhab {
        token == "hanafi" ? .hanafi : .shafi
    }

    private static func highLatitudeRule(from token: String) -> HighLatitudeRule {
        switch token {
        case "automatic": .automatic
        case "middleOfTheNight": .middleOfTheNight
        case "seventhOfTheNight": .seventhOfTheNight
        case "twilightAngle": .twilightAngle
        case "none": .unadjusted
        default: .automatic
        }
    }
}
